### ISSUE TITLE
Problem with Projectile shooting in Phaser and uncertainity of websockets

### ISSUE THREAD

**Issue by GeorgeAlexRob** [rstropek-teaching/real-time-game-quiz-4bhif-AlexGeorgeRob#1]
I'm kinda stuck with the shooting mechanic in my game, it just doesn't want to work like I want it to. The bullets seem to only update once in a while, which admittedly looks pretty funny, but is not the effect I want. (If you want to see it for yourself, clone the repository, run "node server.js" and the head to the browser and type in localhost:65500/index.html; You might have to run gulp first, but I don't think that'll be necessary)

Now I'm kinda out of ideas, I tried a few different methods of the bullet firing, experimented with the velocity variable of the bullet but nothing really helped. I've used Your space shoot example from the coding dojo as a template by the way, it really helped me get into phaser.io. 

My second problem is that I'm not really sure how to implement websockets in this game. I mean an idea would be that I'll send the X and Y coordinates of the players to the server and then the server checks for collisions and everything, but that sounds a bit complicated. I would appreciate any advice regarding this. 

Best regards
Alex Rob

**Comment by rstropek** [comment]
Hi!

No problem, I am happy to help. I will be able to look into the code today in the evening.

Greetings,
Rainer Stropek

**Comment by rstropek** [comment]
Hi @AlexGeorgeRob 

I took a look at your code. I like that you took the challenge to build a phaser.io game. Before I try to give you tips regarding your questions, a more general tip: Before you invite other people to contribute to your code or to give support, make your project as easy to consume as possible. In your case, I recommend adding scripts to *package.json* that can be used to build, start, etc. your project. In school, sending a project with a half-baked configuration is ok - after all, you are still learning. In real life, careful project configuration and good coding style will be your *first impression* to the community and you want to get it right.

Now to your question: As far as I understood your code, your [bullet positioning code](https://github.com/rstropek-teaching/real-time-game-quiz-4bhif-AlexGeorgeRob/blob/ad8ed329411661343fcbd11fdc76aaaa99663b32/client/scripts/app.ts#L153) runs only once and you do not set `bullet.body.velocity`. Therefore, you position the bullet only once. I think you have to initially position the bullet and then set `velocity` according to the orientation of the player's tank.

Regarding Websockets: You game is quite complex if built as a multi-player game. *Phaser.io* offers a lot of features to do collision detection. You could use them to e.g. detect a collision of the bullet and a wall. However, when it comes to two players, where do you want to do the collision detection? On the PC of player 1 or player 2? How do you want to handle latency (e.g. player 2 has moved while your program was busy transmitting Websocket message)?

My tip: Simplify your game. Build a single player game and use Websockets only to transmit e.g. scores. With that, you can use all the nice features of *Phaser.io* and still practice Websockets.

Hope that helps. Greetings,
Rainer Stropek.

**Comment by GeorgeAlexRob** [comment]
Hi!

First of all, thank You for Your elaborate feedback, I'm very thankful for every kind of tips on improving my development skills. I've made two scripts now, one that builds the game, and one that starts the server and opens the game in the browser. Now all one has to do is clone the repository and run `npm run build` and then run `npm start` and everything should work perfectly (Reason I did not join those two together is because I've heard its bad practice to build everything everytime You run the game, I might be wrong though). 

I've worked around my bullet problem by using the `game.physics.arcade.moveToXY()` function and using the _X_ and _Y_ coordinate the tank faces towards. I used this method because `body.velocity` seemed kind of buggy with the bullets. 

I will try to make a singleplayer game out of it, as You suggested, with increasingly more difficult enemies  and a highscore with a name attached to it that gets submitted to the server (And I might make a second page where the top 10 highest scores get saved and printed out in a table).

Again, thank You very much for Your help.

Best regards,
Alex Rob
